File: modules/session/blob.go
package session

import (
	"crypto/sha256"
	"fmt"
	"os"
	"path/filepath"
)

// BlobStore implements content-addressed storage using SHA-256 hashes as keys.
// Each blob is stored as a file named by its hex-encoded SHA-256 hash.
type BlobStore struct {
	dir string
}

// NewBlobStore creates a BlobStore that stores blobs in the given directory.
// The directory is created if it does not exist.
func NewBlobStore(dir string) *BlobStore {
	return &BlobStore{dir: dir}
}

// Put writes data to the store and returns its SHA-256 hex hash.
// If a blob with the same content already exists, it is not written again.
func (b *BlobStore) Put(data []byte) (string, error) {
	hash := sha256.Sum256(data)
	hexHash := fmt.Sprintf("%x", hash)

	path := filepath.Join(b.dir, hexHash)

	// Check if already exists (idempotent)
	if _, err := os.Stat(path); err == nil {
		return hexHash, nil
	}

	if err := os.MkdirAll(b.dir, 0o755); err != nil {
		return "", fmt.Errorf("blob store mkdir: %w", err)
	}

	if err := os.WriteFile(path, data, 0o644); err != nil {
		return "", fmt.Errorf("blob store write: %w", err)
	}

	return hexHash, nil
}

// Get reads a blob by its SHA-256 hex hash.
// Returns an error if the blob does not exist.
func (b *BlobStore) Get(hash string) ([]byte, error) {
	path := filepath.Join(b.dir, hash)
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("blob store get %s: %w", hash, err)
	}
	return data, nil
}

// Has returns true if a blob with the given hash exists in the store.
func (b *BlobStore) Has(hash string) bool {
	path := filepath.Join(b.dir, hash)
	_, err := os.Stat(path)
	return err == nil
}

File: packages/session/src/blob.ts
/**
 * @module session-blob
 *
 * Session blob storage — persistent binary data associated with sessions.
 *
 * Manages storage and retrieval of binary data (screenshots, file snapshots,
 * etc.) that are associated with agent sessions but stored separately from
 * the conversation history.
 */
import { createHash } from "node:crypto";
import { mkdir, readFile, stat, writeFile } from "node:fs/promises";
import { join } from "node:path";
import { OrangeError, newIOError } from "@orangecoding/core";

/** Matches a valid SHA-256 hex hash (64 lowercase hex chars). */
const VALID_HEX_HASH = /^[0-9a-f]{64}$/;

/**
 * BlobStore implements content-addressed storage using SHA-256 hashes as keys.
 * Each blob is stored as a file named by its hex-encoded SHA-256 hash.
 */
export class BlobStore {
  constructor(private readonly dir: string) {}

  /**
   * Put writes data to the store and returns its SHA-256 hex hash.
   * If a blob with the same content already exists, it is not written again.
   */
  async put(data: Uint8Array): Promise<string> {
    const hexHash = createHash("sha256").update(data).digest("hex");
    const path = join(this.dir, hexHash);

    // Check if already exists (idempotent)
    try {
      await stat(path);
      return hexHash;
    } catch {
      // File does not exist; proceed to write.
    }

    try {
      await mkdir(this.dir, { recursive: true });
    } catch (err) {
      throw newIOError(`blob store mkdir: ${(err as Error).message}`);
    }

    try {
      await writeFile(path, data);
    } catch (err) {
      throw newIOError(`blob store write: ${(err as Error).message}`);
    }

    return hexHash;
  }

  /**
   * Get reads a blob by its SHA-256 hex hash.
   * Throws OrangeError if the hash is invalid or the blob does not exist.
   */
  async get(hash: string): Promise<Uint8Array> {
    if (!VALID_HEX_HASH.test(hash)) {
      throw newIOError(`blob store: invalid hash format "${hash}"`);
    }
    const path = join(this.dir, hash);
    try {
      return await readFile(path);
    } catch (err) {
      throw newIOError(`blob store get ${hash}: ${(err as Error).message}`);
    }
  }

  /**
   * Has returns true if a blob with the given hash exists in the store.
   */
  async has(hash: string): Promise<boolean> {
    if (!VALID_HEX_HASH.test(hash)) {
      return false;
    }
    const path = join(this.dir, hash);
    try {
      await stat(path);
      return true;
    } catch {
      return false;
    }
  }
}
